pkg/querier/http.go

```diff
@@ -38,41 +40,53 @@ type QueryResponse struct {
 	Result     parser.Value     `json:"result"`
 }
 
-type Engine interface {
-	Query(logql.Params) logql.Query
-}
-
 // nolint // QuerierAPI defines HTTP handler functions for the querier.
 type QuerierAPI struct {
-	querier Querier
-	cfg     Config
-	limits  querier_limits.Limits
-	engine  Engine
+	querier  Querier
+	cfg      Config
+	limits   querier_limits.Limits
+	engineV1 logql.Engine // Loki's current query engine
+	engineV2 logql.Engine // Loki's next generation query engine
+	logger   log.Logger
 }
 
 // NewQuerierAPI returns an instance of the QuerierAPI.
-func NewQuerierAPI(cfg Config, querier Querier, limits querier_limits.Limits, logger log.Logger) *QuerierAPI {
-	engine := logql.NewEngine(cfg.Engine, querier, limits, logger)
+func NewQuerierAPI(cfg Config, querier Querier, limits querier_limits.Limits, metastore metastore.Metastore, logger log.Logger) *QuerierAPI {
 	return &QuerierAPI{
-		cfg:     cfg,
-		limits:  limits,
-		querier: querier,
-		engine:  engine,
+		cfg:      cfg,
+		limits:   limits,
+		querier:  querier,
+		engineV1: logql.NewEngine(cfg.Engine, querier, limits, logger),
+		engineV2: engine.New(cfg.Engine, metastore, limits, logger),
+		logger:   logger,
 	}
 }
 
 // RangeQueryHandler is a http.HandlerFunc for range queries and legacy log queries
 func (q *QuerierAPI) RangeQueryHandler(ctx context.Context, req *queryrange.LokiRequest) (logqlmodel.Result, error) {
+	var result logqlmodel.Result
+	logger := utillog.WithContext(ctx, q.logger)
+
 	if err := q.validateMaxEntriesLimits(ctx, req.Plan.AST, req.Limit); err != nil {
-		return logqlmodel.Result{}, err
+		return result, err
 	}
 
 	params, err := queryrange.ParamsFromRequest(req)
 	if err != nil {
-		return logqlmodel.Result{}, err
+		return result, err
+	}
+
+	if q.cfg.Engine.EnableLQE {
+		query := q.engineV2.Query(params)
+		result, err = query.Exec(ctx)
+		if !errors.Is(err, engine.ErrNotSupported) {
+			level.Error(logger).Log("msg", "query execution failed with new query engine", "err", err)
+			return result, errors.Wrap(err, "failed with new execution engine")
+		}
```

Comment:
This branch would run even if err is nil, is that what you intended?

Reply:
No, that was not intended. Good catch!